### ISSUE TITLE
Add support to run schematics that are part of custom plugins within an existing workspace.

### ISSUE THREAD

**Issue by niki4810** [nrwl/nx#3231]
## Description

We can currently create custom schematics using the `nx generate @nrwl/workspace-schematic` command and run them locally within an existing workspace (consisting of apps and libs) via `nx workspace-schematic <schematic-name>`

However, when creating a custom plugin via `nx generate @nrwl/nx-plugin:plugin` and creating a schematic under `libs/<plugin-name>/src/schematics`, there is currently no way to run this plugin locally to use/test our schematics (unless we publish)

- Here is a sample project to illustrate the issue: https://github.com/niki4810/myorg
- You can run the schematic in `tools`: https://github.com/niki4810/myorg/tree/master/tools/schematics/custom-react
- But currently, there is no way to run the same schematic that is part of the plugin: https://github.com/niki4810/myorg/tree/master/libs/custom-react/src/schematics/custom-react

Please refer to this slack thread for more details: https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1592940422065400

## Motivation
Having the ability to run plugins (with custom schematics) helps in:
- following the same syntax external plugins: `nx generate <plugin-name>:<schematic>` instead of new command `nx workspace-schematic <schematic>`
- Have the ability to list the schematics via: `nx list <plugin-name>` 

## Suggested Implementaion
- I would expect it to work the same way as any published community plugins work, i.e. `nx generate <plugin-name>:<schematic>`

## Alternate Implementations
- N/A

**Comment by webberwang** [comment]
Is there a way to yarn link without having to publish?

**Comment by pascalbe-dev** [comment]
This is actually possible. You can build the plugin locally, so it ends up in your dist folder. Then you can run it directly from that build target location via nx. In your sample repo, it would work like that:

`
nx g ./dist/libs/custom-react:library
`

Ofc you can also pass options.

**Comment by Carniatto** [comment]
This is not really the ideal solution, cuz if anyone in the team wants to use it they'd need to build it every time. Or persist the dist folder

**Comment by Hotell** [comment]
this PR adds similar DX capabilities to workspace-schematic, like nx plugins provide
https://github.com/nrwl/nx/pull/3617

**Comment by mehrad-rafigh** [comment]
This only works if you have one generator that does not depend on another generator inside the plugin 👎🏽

**Comment by hadrien-toma** [comment]
I also supposed it was working like for schematics and finally ended here. I would be very happy to be to use this really well defined enhancement.

**Comment by koliveira15** [comment]
This would help our org immensely. We have many workspace-generators that are being converted to plugins so other nx repos within our org can consume. Building the plugin before executing within the same workspace the plugin exists is not a great DX.

**Comment by AgentEnder** [comment]
This is something we are re-evaluating post Nx 13.0, as we agree that it would help DX tremendously.

As a quick note, there are some workarounds or other things you can do to enhance DX with the current setup.

### Workspace Generator -> Generators in a Plugin
If moving a workspace generator into a plugin, you can have the existing workspace-generator import the plugins generator and call it. This way people don't have to switch the commands they are used to using already either.

### Executors from a local nx plugin
For executors in a "workspace-plugin", you can make sure projects that use the plugin have an implicit dependency on the plugin project. This ensures that the plugin is built (and in its latest state) before the executor runs. Since nx caches these builds, it shouldn't extend your command run times.

Just adding the dependency like above, is all thats needed for build targets. This is because of the targetDependencies section of nx.json's default contents. If your executor has a generic targetName, you would need to add a dependsOn block to the target, and may want to add a target on the plugin called "build-plugin" or similar, to avoid building non plugin projects before the executor runs. 

For example:

If you have a plugin that contains a new serve executor, and a lint executor, you may have a `workspace.json` with projects such as below:
```json
{
  "projects": {
    "my-plugin": {
      "targets": {
        "build": {
          "executor": "@nrwl/node:package"
        },
        "build-plugin": {
          "executor": "@nrwl/run-commands",
          "options": {
            "command": "nx build my-plugin"
          }
        }
      }
    },
    "my-app": {
      "targets": {
        "serve": {
          "executor": "./dist/packages/my-plugin:serve",
          "dependsOn": [
            {
              "target": "build-plugin",
              "projects": "dependencies"
            }
          ]
        },
        "lint": {
          "executor": "./dist/packages/my-plugin:lint",
          "dependsOn": [
            {
              "target": "build-plugin",
              "projects": "dependencies"
            }
          ]
        }
      }
    }
  }
}
```

If multiple projects use those custom executors in your workspace, it would make more sense to remove the dependsOn blocks, and use targetDependencies in `nx.json`. That would look something like:

```json
{
  "targetDependencies": {
    "serve": [
      {
        "target": "build-plugin",
        "projects": "dependencies"
      }
    ],
    "lint": [
      {
        "target": "build-plugin",
        "projects": "dependencies"
      }
    ]
  }
}
```

**Comment by gioragutt** [comment]
Any update on that happening?

**Comment by AgentEnder** [comment]
This is still something that I'd like to get back to implementing, but don't have the capacity at the current moment.

**Comment by gioragutt** [comment]
@AgentEnder From what you can currently estimate, how much would such a task take? And how much longer would it take for a non-core contributor?

**Comment by AgentEnder** [comment]
I don't think it would be a tremendous lift to get working, but there would be a lot of cleanup work to go with it. The way I would currently approach the problem (assuming the others on the team agree w/ the approach), would look something like this:

1. Update logic for generate (`nx g ...`) to use an extended resolver instead of just require.resolve. The extension would look over the files in the repo, looking for package.json files (excluding the root). If the collection name matched any of those package.json's name, we could match it up that way. 
2. Update logic for run-one (`nx run ...`) to use the same resolver
3. Add a migration to create a project in the workspace's tools directory if they currently have workspace generators. Setup a minimal package.json and such in it, so that the custom logic in `nx workspace-generator` isn't needed anymore
4. Deprecate `nx workspace-generator`, and update the existing logic to just forward into `nx g` with the new project in the tools dir as the collection.

The really cool part of this is that it also greatly simplifies the process for writing and maintaining custom executors, and the setup for workspace-generators / custom executors becomes the same as the setup for a new Nx plugin. They could be published if desired.

A point of complication comes from the generators + executors needing to be built first. When running `nx g @company/my-workspace-plugin:app`, we would have to make sure that the dist folder for my-workspace-plugin matches the hash for the project and rebuild before generating otherwise. The same is true for executors. I can't speak to how hard this would be, I've not personally worked on a portion that impacts things like this.

The only reason this would be harder for a non-core contributor would be that its going to require an agreement in _how_ we think it should work, and there is a pretty deep knowledge requirement for updating certain areas of the CLI

**Comment by gioragutt** [comment]
@AgentEnder awesome man, really cool to hear about the thought process for something like that.

For now, I've been able to overcome my issues by having a generator built with `workspace-generator`, and for the executors I did the following:

1. Created a plug-in with an executor.
1. Added a target that builds the plug-in using 'run-commands'
1. Added a target to another lib that uses this executor with a 'dist' path
1. Added a 'dependsOn' section to that target that runs the target from step #2 (with projects: "dependencies")
1. Added the plug-in to the lib's implicit dependencies.

This works, but has the side effect of making the plug-in appear in the built lib's dependencies (as there's no "implicitDevDependencies").

For that issue I created another executor, that wraps "@nrwl/node:package" which removes the plug-in from the generated package.json's dependencies.

It's really cool that it works, but having it, you know, Just Work™️, would be an amazing developer experience.

**Comment by AgentEnder** [comment]
Yeah definitely. The implicitDeps + target dependencies is definitely my recommendation, and is the current best DX. I mostly work on web apps, hadn't thought about the generatePackageJson causing that. Perhaps opening another issue for implicit dependencies to be ignored by generate package json could get handled quicker?

**Comment by gioragutt** [comment]
@AgentEnder

> Perhaps opening another issue for implicit dependencies to be ignored by generate package json could get handled quicker?

Yes, I can definitely do that.

**Comment by gioragutt** [comment]
@AgentEnder ☝🏻 LMK if something else is needed

**Comment by Person-93** [comment]
I'm working on a plugin that has a generator, several executors, and I plan to add a graph extender plugin as well.
The generator is supposed to make projects that use those executors and rely on the graph extender to find dependencies.
Is there a way to use the plugin in the workspace while I'm developing it?

**Comment by gioragutt** [comment]
@Person-93 see [this](https://github.com/nrwl/nx/issues/3231#issuecomment-922090762) comment from above.

Currently, you'd have to link the executors to the `dist` folder, this will solve the executor problem.
For generators, you can use the `workspace-generator` feature.

**Comment by Person-93** [comment]
@gioragutt I haven't tried the executors yet, but I haven't been able to get the generator working, Here's what I've done so far.

### Make Workspace Generator
I ran the generator generator. This put the `index.ts` and `schem.json` in a new generator in the tools directory. All is well.

### Get the schema
I first tried using a `$ref` but it couldn't resolve the file.

Next I tried using a relative symlink to the plugin generator's schema, but it just copied the symlink into the dist directory as is and it no longer pointed to anything.

For now, I just copied the schema file, If all else works, it shouldn't be too difficult to write a script to keep it in sync automatically, but I'd prefer not to have to do that as it makes the repository structure more complex.

### Import the Plugin Generator
I had to remove the `rootDir` option from `tools.tsconfig.json`. I think that this caused my next problem.

I imported the plugin and its schema declaration into the workspace generator. My IDE insists that it can't find the module but it does compile, so I guess that's a problem for another day.

It compiles but it doesn't run. The error message is that it can't find the module in the tools directory. I looked in `dist/out-tsc/tools` and it has a generators directory which seems to be a copy of `tools/generators` from outside the dist directory. It has a a libs directory which seems to be the compiled output from the imported plugin generator and it has a nested `tools/generators` directory which has the compiled output from the workspace generator.

**TLDR:** I haven't got it working.

**Comment by gioragutt** [comment]
> I first tried using a $ref but it couldn't resolve the file.

I think I'm missing some context as to what you're trying to achieve.
Generating with Workspace Generators works out of the box, so what are you trying to do other than that?

**Comment by gioragutt** [comment]
@Person-93 

> I first tried using a $ref but it couldn't resolve the file.

I think I'm missing some context as to what you're trying to achieve.
Generating with Workspace Generators works out of the box, so what are you trying to do other than that?

**Comment by Person-93** [comment]
@gioragutt sorry, let me backtrack a few steps.

I'm working on a plugin that I intend to publish and use in other workspaces. I'd like to use that plugin in the workspace that it lives in.

The comment you linked above suggested importing the plugin's generator into a workspace generator.

> ### Workspace Generator -> Generators in a Plugin
> 
> If moving a workspace generator into a plugin, you can have the existing workspace-generator import the plugins generator and call it. This way people don't have to switch the commands they are used to using already either.

That's what I was trying to do except there was no pre-existing workspace generator. I made a workspace generator specifically to call the plugin's generator.

As for the `$ref`, I tried defining the workspace generator's schema like this:
```json
{
  "$schema": "http://json-schema.org/schema",
  "$ref": "path/to/plugin/generator.schema"
}
```
I was hoping to avoid duplicating the schema from the generator's plugin and having to keep it in sync.

**Comment by gioragutt** [comment]
@Person-93 gotcha. Unfortunately, I don't have an answer for the json schema problem. When I created a generator based on `@nrwl/node:library` I just copied the schema from node_modules and altered it as needed.

@AgentEnder perhaps you have some insights as to how to manage json schemas of generators that use other generators, beyond copy-pasting and altering the original schema.

**Comment by gioragutt** [comment]
@Person-93 I thought of a good idea for a tool, I'll share it here so maybe someone will take the time to implement it -

Create a tool that would convert the `.d.ts` schema to a json schema. It would use the declared type definitions, as well as comments, to create the json schema.

Perhaps it can be part of NX so that only `.d.ts` schemas would be required 🙄

**Comment by Person-93** [comment]
@gioragutt , thanks for your quick replies and general helpfulness. :)

-----

The schema thing isn't the main issue. While it's not ideal, I can automate a way to keep it in sync.

I haven't been able to run the generator at all. Importing the plugin's generator into a workspace generator has not worked.

> ### Import the Plugin Generator
> 
> I had to remove the `rootDir` option from `tools.tsconfig.json`. I think that this caused my next problem.
> 
> I imported the plugin and its schema declaration into the workspace generator. My IDE insists that it can't find the module but it does compile, so I guess that's a problem for another day.
> 
> It compiles but it doesn't run. The error message is that it can't find the module in the tools directory. I looked in `dist/out-tsc/tools` and it has a generators directory which seems to be a copy of `tools/generators` from outside the dist directory. It has a a libs directory which seems to be the compiled output from the imported plugin generator and it has a nested `tools/generators` directory which has the compiled output from the workspace generator.
> 
> **TLDR:** I haven't got it working.

-----


> @Person-93 I thought of a good idea for a tool, I'll share it here so maybe someone will take the time to implement it -
> 
> Create a tool that would convert the `.d.ts` schema to a json schema. It would use the declared type definitions, as well as comments, to create the json schema.
> 
> Perhaps it can be part of NX so that only `.d.ts` schemas would be required roll_eyes

That's an awesome idea! i thought about something like that, but I wasn't sure how to make sure the `x-prompt` property gets included. For a simple straightforward declaration, it could look for a special directive in the comment. But for a type that is a union or intersection of other types, especially if it uses `Partial` and/or `Omit`, it gets a bit tricky.

**Comment by gioragutt** [comment]
@Person-93 oh, so you're trying to run a generator that's defined as a plugin library, inside of a workspace generator. That's indeed a tricky situation, and I don't think I have an idea here 😔

**Comment by JoA-MoS** [comment]
I thought I would offer this approach to see what people thought and maybe give some others an alternate approach. Some additional features this method has that some others may not:

### Features
1. **Auto registers plugins after `npm ci` or `npm install`** - The intent here is to simplify new team member onboarding using the plugin without a bunch of manual steps
2. **Plugins are available in NX Console** - Plugins can be executed using the NX console. This also helps those who primarily use the NX Console but also helps in development as you can preview how it works in the NX Console

### Approach

1. Add a task to your plugins `project.json` to link the plugin
```
 "link-plugin": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "commands": [
          "npm link ./dist/packages/nx-tools",
          "npm link @my-org/nx-tools"
        ],
        "parallel": false
      }
 ```
 2. Update the nx.json or project.json to setup a dependency between link-plugin and build. I am showing nx.json here
 ```
 "targetDependencies": {
    "link-plugin": [
      {
        "target": "build",
        "projects": "self"
      }
    ]
  },
  ```
  3. Add a postinstall script to your npm scripts that will run any link-plugin task in your workspace
  ```
  "scripts": {
    "register-local-nx-plugins": "nx run-many --target=link-plugin --all",
    "postinstall": "npm run register-local-nx-plugins"
  },
  ```
  
  > I did not add an unlink-plugin task as it didn't really serve a purpose in my usecase
  
  Let me know if there are any thoughts on this method
